pkg/operator/ceph/object/shared_pools.go

```diff
@@ -0,0 +1,695 @@
+	// if zoneChanged || zoneGroupChanged {
+	// 	err = commitConfigChanges(objContext)
+	// 	if err != nil {
+	// 		return err
+	// 	}
+	// }
+
+	logger.Infof("Successfully configured RADOS namespaces for object store %q", objContext.Name)
+	return nil
+}
+
+func adjustZoneDefaultPools(zone ZoneJSON, spec cephv1.ObjectSharedPoolsSpec) (bool, error) {
+	defaultMetaPool := getDefaultMetadataPool(spec)
+	if defaultMetaPool == "" {
+		// default pool is not presented in shared pool spec
+		return false, nil
+	}
+	name, err := zone.getString("name")
+	if err != nil {
+		return false, fmt.Errorf("unable to get zone name: %w", err)
+	}
+	// add zone namespace to metadata pool to safely share accorss rgw instances or zones.
+	// in non-multisite case zone name equals to rgw instance name
+	defaultMetaPool = defaultMetaPool + ":" + name
+	zonePoolNSSuffix := map[string]string{
+		"domain_root":     ".meta.pool",
+		"control_pool":    ".control",
+		"gc_pool":         ".log.gc",
+		"lc_pool":         ".log.lc",
+		"log_pool":        ".log",
+		"intent_log_pool": ".log.intent",
+		"usage_log_pool":  ".log.usage",
+		"roles_pool":      ".meta.roles",
+		"reshard_pool":    ".log.reshard",
+		"user_keys_pool":  ".meta.users.keys",
+		"user_email_pool": ".meta.users.email",
+		"user_swift_pool": ".meta.users.swift",
+		"user_uid_pool":   ".meta.users.uid",
+		"otp_pool":        ".otp",
+		"notif_pool":      ".log.notif",
+	}
+	hasChanged := false
+	for pool, nsSuffix := range zonePoolNSSuffix {
+		// replace rgw internal index pools with namespaced metadata pool
+		prev, err := zone.getString(pool)
+		if err != nil {
+			return false, fmt.Errorf("unable to get pool %s for zone %s: %w", pool, name, err)
+		}
+		updated := defaultMetaPool + nsSuffix
+		if prev == updated {
+			continue
+		}
+		logger.Debugf("update shared pool %s for zone %s: %s -> %s", pool, name, prev, updated)
+		zone.setString(pool, updated)
+		hasChanged = true
+	}
+	return hasChanged, nil
+}
+
+func adjustZonePlacementPools(zone ZoneJSON, spec cephv1.ObjectSharedPoolsSpec) (bool, error) {
+	name, err := zone.getString("name")
+	if err != nil {
+		return false, fmt.Errorf("unable to get zone name: %w", err)
+	}
+
+	old, err := zone.getPoolPlacements()
+	if err != nil {
+		return false, fmt.Errorf("unable to get pool placements for zone %s: %w", name, err)
+	}
+	oldMap := make(map[string]ZonePlacementPool, len(old))
+	for _, pp := range old {
+		oldMap[pp.Key] = pp
+	}
+	// map spec to new zone pools config
+	updated := toZonePlacementPools(spec, name)
+	newMap := make(map[string]ZonePlacementPool, len(updated))
+	for i, pp := range updated {
+		if oldPl, ok := oldMap[pp.Key]; ok {
+			// preserve original values which are not covered by pool placement CRD
+			pp.Val.IndexType = oldPl.Val.IndexType
+			pp.Val.InlineData = oldPl.Val.InlineData
+			updated[i].Val.IndexType = oldPl.Val.IndexType
+			updated[i].Val.InlineData = oldPl.Val.InlineData
+		}
+		newMap[pp.Key] = pp
+	}
```

Comment:
I'd like to understand more about what this code's purpose is. I have some concerns, but I might be missing some context.

Preserving original fields could be a source of configuration bugs, possibly if users decide to remove pool placements after they are configured. Generally, things work best when Rook builds up the list of what the config should look like from scratch each time to apply it, but we can deviate from that if we do it on purpose.

It might be easiest if you test manually to try to remove a configured placement and see if it gets removed from Ceph's config.

Let's also discuss with @travisn in huddle what behavior we'd like for users. In my experience, users are sometimes upset when they remove things from a spec and Rook doesn't remove them from the config. However, in this context, doing so could result in data loss or data being made inaccessible, which could be another issue. Likely, very few users will run into either issue, but I'd like to at least have an understanding about what it is and what we intend so we can respond to them if this comes up.